ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFDateDiff.java

```diff
@@ -64,121 +66,46 @@
-    case STRING:
-    case VARCHAR:
-    case CHAR: {
-      String dateString = converter.convert(argument.get()).toString();
-      Date date = DateParser.parseDate(dateString);
-      if (date != null) {
-        return date;
-      }
-      Timestamp ts = PrimitiveObjectInspectorUtils.getTimestampFromString(dateString);
-      if (ts != null) {
-        return Date.ofEpochMilli(ts.toEpochMilli());
-      }
+    if (ts1 == null || ts2 == null) {
       return null;
     }
-    case TIMESTAMP:
-      Timestamp ts = ((TimestampWritableV2) converter.convert(argument.get()))
-        .getTimestamp();
-      return Date.ofEpochMilli(ts.toEpochMilli());
-    case DATE:
-      DateWritableV2 dw = (DateWritableV2) converter.convert(argument.get());
-      return dw.get();
-    case TIMESTAMPLOCALTZ:
-      TimestampTZ tsz = ((TimestampLocalTZWritable) converter.convert(argument.get()))
-          .getTimestampTZ();
-      return Date.ofEpochMilli(tsz.getEpochSecond() * 1000l);
-    default:
-      throw new UDFArgumentException(
-        "TO_DATE() only takes STRING/TIMESTAMP/TIMESTAMPLOCALTZ types, got " + inputType);
-    }
-  }
 
-  private Converter checkArguments(ObjectInspector[] arguments, int i) throws UDFArgumentException {
-    if (arguments[i].getCategory() != ObjectInspector.Category.PRIMITIVE) {
-      throw new UDFArgumentTypeException(0,
-        "Only primitive type arguments are accepted but "
-        + arguments[i].getTypeName() + " is passed. as first arguments");
-    }
-    final PrimitiveCategory inputType =
-        ((PrimitiveObjectInspector) arguments[i]).getPrimitiveCategory();
-    switch (inputType) {
-    case STRING:
-    case VARCHAR:
-    case CHAR:
-      return ObjectInspectorConverters.getConverter(arguments[i],
-        PrimitiveObjectInspectorFactory.writableStringObjectInspector);
-    case TIMESTAMP:
-      return new TimestampConverter((PrimitiveObjectInspector) arguments[i],
-        PrimitiveObjectInspectorFactory.writableTimestampObjectInspector);
-    case TIMESTAMPLOCALTZ:
-      return new PrimitiveObjectInspectorConverter.TimestampLocalTZConverter(
-          (PrimitiveObjectInspector) arguments[i],
-          PrimitiveObjectInspectorFactory.writableTimestampTZObjectInspector
-      );
-    case DATE:
-      return ObjectInspectorConverters.getConverter(arguments[i],
-        PrimitiveObjectInspectorFactory.writableDateObjectInspector);
-    default:
-      throw new UDFArgumentException(
-          " DATEDIFF() only takes STRING/TIMESTAMP/DATEWRITABLE/TIMESTAMPLOCALTZ types as " + (i + 1)
-              + "-th argument, got " + inputType);
-    }
-  }
+    Date date1 = Date.ofEpochMilli(ts1.toEpochMilli());
+    Date date2 = Date.ofEpochMilli(ts2.toEpochMilli());
 
-  private IntWritable evaluate(Date date, Date date2) {
-
-    if (date == null || date2 == null) {
-      return null;
-    }
+    output.set(DateWritableV2.dateToDays(date1) - DateWritableV2.dateToDays(date2));
+    return output;
```

Comment:
Do we need to transform `Timestamp` to `Date` in order to get the days difference. Inside timestamp there is `LocalDateTime` so can't we use the new Java APIs (such as `Duration` or `ChronoUnit`) to get the difference?

Reply:
You are right we can use java APIs  Since we already have functions to convert tiimestamp to date. I am just reusing the existing code instead of adding new.